library/bignum_new.c

```diff
@@ -0,0 +1,129 @@
+ *  not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *  http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ *  WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+#include "common.h"
+
+#if defined(MBEDTLS_BIGNUM_C)
+
+#include "mbedtls/bignum.h"
+#include "bignum_core.h"
+#include "bn_mul.h"
+
+#include <string.h>
+
+#define ciL    (sizeof(mbedtls_mpi_uint))       /* chars in limb  */
+#define biL    (ciL << 3)                       /* bits  in limb  */
+#define biH    (ciL << 2)                       /* half limb size */
+
+void mbedtls_mpi_core_montmul( mbedtls_mpi_uint *X,
+                               const mbedtls_mpi_uint *A,
+                               const mbedtls_mpi_uint *B,
+                               size_t B_len,
+                               const mbedtls_mpi_uint *N,
+                               size_t n,
+                               mbedtls_mpi_uint mm,
+                               mbedtls_mpi_uint *T )
+{
+    memset( T, 0, (2*n+1)*ciL );
+
+    for( size_t i = 0; i < n; i++, T++ )
+    {
+        mbedtls_mpi_uint u0, u1;
+        /* T = (T + u0*B + u1*N) / 2^biL */
+        u0 = A[i];
+        u1 = ( T[0] + u0 * B[0] ) * mm;
+
+        (void) mbedtls_mpi_core_mla( T, n + 2, B, B_len, u0 );
+        (void) mbedtls_mpi_core_mla( T, n + 2, N, n, u1 );
+    }
+
+    mbedtls_mpi_uint carry, borrow;
+
+    carry  = T[n];
+    borrow = mbedtls_mpi_core_sub( X, T, N, n );
+    (void) mbedtls_mpi_core_add_if( X, N, n, ( carry < borrow ) );
+}
+
+mbedtls_mpi_uint mbedtls_mpi_core_mla( mbedtls_mpi_uint *d, size_t d_len,
+                                       const mbedtls_mpi_uint *s, size_t s_len,
+                                       mbedtls_mpi_uint b )
+{
+    mbedtls_mpi_uint c = 0; /* carry */
+    if( d_len < s_len )
+        s_len = d_len;
+    size_t excess_len = d_len - s_len;
+    size_t steps_x8 = s_len / 8;
+    size_t steps_x1 = s_len & 7;
+
+    while( steps_x8-- )
+    {
+        MULADDC_X8_INIT
+        MULADDC_X8_CORE
+        MULADDC_X8_STOP
+    }
+
+    while( steps_x1-- )
+    {
+        MULADDC_X1_INIT
+        MULADDC_X1_CORE
+        MULADDC_X1_STOP
+    }
+
+    while( excess_len-- )
+    {
+        *d += c; c = ( *d < c ); d++;
+    }
+
+    return( c );
+}
+
+mbedtls_mpi_uint mbedtls_mpi_core_sub( mbedtls_mpi_uint *d,
+                                       const mbedtls_mpi_uint *l,
+                                       const mbedtls_mpi_uint *r,
+                                       size_t n )
+{
+    mbedtls_mpi_uint c = 0, t, z;
+
+    for( size_t i = 0; i < n; i++ )
+    {
+        z = ( l[i] <  c );    t = l[i] - c;
```

Comment:
Unnecessary whitespace in
`z = ( l[i] <  c );{HERE} t = l[i] - c;`

Reply:
Yes, these were all brought in as-is from the prototype; I wasn't sure whether to

1. fix them as I brought them in (not done);
2. bring them in and fix them in a later commit (not done);
3. bring them in as-is and leave them (which is what I've done).

if you have a preference for 1 or 2 please let me know

Reply:
Actually, re-reviewing these, these are all part of alignment of more than one statement on a line, which is strictly against the coding standard.

However, as these breaches of the standard are made in order to improve the readability of the code, I think they should remain.

Reply:
In general I have a preference for 2. Regarding what the fix should be in this specific instance, I don't have a strong opinion.

Reply:
I think I am going to leave style breaches such as this - made in the original for readability - as they are. For things like the subsequent one - `(2*n+1)*ciL` which clearly isn't an aid to readability - I will update in a "fix style"-type commit

Reply:
Although both `t` and `z` is local to the loop and aren't used across iterations, so they could be both declared and initialised here and then both coding standard and readability would suggest that they should be in their individual lines.

Also, it is subjective, but I find it more readable if only one statement is in one line. Same goes for the next line (line 107) as well.

Reply:
putting the declarations into the loop body makes it clearly less readable

```
    for( size_t i = 0; i < n; i++ )
    {
        mbedtls_mpi_uint z = ( l[i] < c );
        mbedtls_mpi_uint t = l[i] - c;
        c = ( t < r[i] ) + z;
        d[i] = t - r[i];
    }
```